cytosnake/utils/config_utils.py

```diff
@@ -131,3 +132,73 @@ def load_cytosnake_configs() -> dict:
     # gets absolute path to cytosnake configs and load the configs
     cytosnake_config_path = cyto_paths.get_cytosnake_config_path()
     return load_configs(cytosnake_config_path)
+
+
+def update_config(
+    config_file_path: str | pathlib.Path,
+    new_key: str,
+    new_value: str | Any,
+    update=False,
+) -> None:
+    """This updates config level in the upper level.
+
+    Parameters
+    ----------
+    key : str
+        key to add into the config
+    value : str | Number
+        Value to add to the given key.
+    update : bool
+        Update value to existing key
+
+    Raises
+    ------
+    ValueError
+        raised if the key value cannot be converted into a string
+    TypeError
+        raised if there is a type error with any parameter
+    FileNotFoundError
+        raised if the config file is not found
+    KeyError
+        raised if given key exists within the config file
+    """
+
+    # type checking
+    if not isinstance(config_file_path, pathlib.Path):
+        if isinstance(config_file_path, str):
+            config_file_path = pathlib.Path(config_file_path).resolve(strict=True)
+        raise TypeError(
+            "'config_file_path' must be str or pathlib.Path."
+            f"not: {type(config_file_path)}"
+        )
+    if not isinstance(new_key, str):
+        try:
+            new_key = str(new_key)
+        except Exception:
+            raise ValueError("Unable to convert key value into a string")
+    if not isinstance(new_value, (str, float, int)):
+        raise TypeError("value must either be a string, float, or int")
+
+    # open config file
+    loaded_configs = load_configs(config_path=config_file_path)
+
+    # updating configs
+    new_key_value_pair = {new_key: new_value}
+    if new_key in [keys for keys in loaded_configs.keys()]:
+        if update:
+            loaded_configs.update(new_key_value_pair)
+        raise KeyError(
+            f"key: `{new_key}` already exists in the config. "
+            "Please use a different key to prevent key clashing"
+        )
+    loaded_configs.update(new_key_value_pair)
```

Comment:
Double checking here: it looks like there may be scenario for `loaded_configs.update(new_key_value_pair)` to run twice. Is this intentional? If not, consider revising this block to help ensure there's no unnecessary repetition of work involved.

Reply:
Hey Dave! The intentional code block prevents overwriting values associated with existing keys. It first checks if the key is already present in the config file. If found, it evaluates the 'update' flag to determine if you want to modify the value. If 'update' is not set to True, it raises an error to avoid unintentional changes. However, if the key is not found in the config file, the code will automatically add it.

Reply:
Thanks @axiomcura , and apologies, I think I may still not understand and might not have been specific enough with my message. It looks like `loaded_configs.update(new_key_value_pair)` runs twice and under different conditions. The conditional check doesn't necessarily prevent the `new_key_value_pair` from being updated in the configs (the 2nd `loaded_configs.update(new_key_value_pair)` outside of the if-check). I asked to make sure this didn't need to be placed within an `else` or `elif`, and generally to deduplicate the code / practice DRY.